### PR TITLE
Extracted actor Voices from Selectable into own Voiced trait

### PR DESCRIPTION
**Pull request by reaperrr**
And implemented `IVoiced` interface to move `Voiced` to Common.

### CODE REVIEW

**Comment by abcdefg30** [comment]
> • track down and fix the issue that Select voice is played twice when actor(s) are selected via group key.

Would fix #5770 then.

**Comment by Mailaender** [comment]
How about calling it `Voiced`?

File: OpenRA.Game/WorldUtils.cs [hunk 1]
```diff
@@ -44,8 +44,8 @@ public static IEnumerable<Actor> FindActorsInCircle(this World world, WPos origi
 
 		public static bool HasVoices(this Actor a)
 		{
-			var selectable = a.Info.Traits.GetOrDefault<SelectableInfo>();
-			return selectable != null && selectable.Voice != null;
+			var hasVoices = a.HasTrait<IWithVoices>();
+			return hasVoices && a.TraitOrDefault<IWithVoices>().VoiceSet != null;
```
Comment (abcdefg30) [inline OpenRA.Game/WorldUtils.cs]: What about something like
```c#
var voice = a.TraitOrDefault<IWithVoices>();
return voice != null && voice.VoiceSet != null;

File: OpenRA.Game/WorldUtils.cs [hunk 2]
```diff
@@ -56,9 +56,11 @@ public static bool HasVoice(this Actor a, string voice)
...
 		{
-			var selectable = a.Info.Traits.GetOrDefault<SelectableInfo>();
-			if (selectable == null) return null;
-			var v = selectable.Voice;
+			var hasVoices = a.HasTrait<IWithVoices>();
+			if (!hasVoices)
+				return null;
+
+			var v = a.TraitOrDefault<IWithVoices>().VoiceSet;
 			return (v == null) ? null : a.World.Map.Rules.Voices[v.ToLowerInvariant()];
```
Comment (abcdefg30) [inline OpenRA.Game/WorldUtils.cs]: Same question here.

File: OpenRA.Game/WorldUtils.cs (hunk 1 above)
Comment (Mailaender) [inline OpenRA.Game/WorldUtils.cs]: Saves one trait lookup.

Comment (Phrohdoh) [inline OpenRA.Game/WorldUtils.cs]: `TraitOrDefault<T>` doesn't make sense if the actor `HasTrait<T>`.

**Comment by Phrohdoh** [comment]
AFAIK With* is typically reserved for rendering traits.

File: OpenRA.Game/WorldUtils.cs (hunk 1 above)
Comment (penev92) [inline OpenRA.Game/WorldUtils.cs]: What @abcdefg30 said.

**Comment by Mailaender** [comment]
I think you can also deprecate the weird

```yaml
Selectable:
    Selectable: false
```

setting then.

**Comment by reaperrr** [comment]
@Mailaender: That setting is weird indeed, but I want to deal with Selectable in a later, dedicated PR anyway (separating selection bounds from visible selection box), so I don't want to touch that in this PR.

**Comment by pchote** [comment]
IIRC that Selectable property was there as a workaround for #2922, which has now been fixed.

**Comment by abcdefg30** [comment]
Needs a rebase.

File: OpenRA.Game/WorldUtils.cs [hunk 3]
```diff
@@ -56,9 +56,7 @@ public static bool HasVoice(this Actor a, string voice)
 
 		public static SoundInfo GetVoices(this Actor a)
 		{
-			var selectable = a.Info.Traits.GetOrDefault<SelectableInfo>();
-			if (selectable == null) return null;
-			var v = selectable.Voice;
+			var v = a.TraitOrDefault<IVoiced>().VoiceSet;
```
Comment (penev92) [inline OpenRA.Game/WorldUtils.cs]: You need to keep the null check here because `TraitOrDefault` can give you null, and you have no guarantees that whoever calls this method will check beforehand.

File: mods/ts/rules/civilian-infantry.yaml [hunk 4]
```diff
@@ -284,8 +293,8 @@ DOGGIE:
 		Range: 4c0
 	Mobile:
 		Speed: 113
-	Selectable:
-		Voice: Fiend
+	Voiced:
+		VoiceSet: Fiend
```
Comment (penev92) [inline mods/ts/rules/civilian-infantry.yaml]: Why is this no longer selectable?

File: mods/ts/rules/civilian-infantry.yaml [hunk 5]
```diff
@@ -314,8 +323,8 @@ VISSML:
 		Speed: 113
 		ROT: 16
 	-Crushable:
-	Selectable:
-		Voice: Fiend
+	Voiced:
+		VoiceSet: Fiend
```
Comment (penev92) [inline mods/ts/rules/civilian-infantry.yaml]: And this?

**Comment by penev92** [comment]
Changes look OK. Will test ingame when I can.

**Comment by reaperrr** [comment]
Updated (reimplemented null check in WorldUtils.GetVoices) and rebased.

File: OpenRA.Game/Sound/Sound.cs [hunk 6]
```diff
@@ -364,30 +364,30 @@ public static float VideoSeekPosition
 			return true;
 		}
 
-		public static bool PlayVoice(string phrase, Actor voicedUnit, string variant)
+		public static bool PlayVoice(string phrase, Actor voicedActor, string variant)
```
Comment (pchote) [inline OpenRA.Game/Sound/Sound.cs]: It would be great if we could move this into the IVoiced trait implementations themselves.  The sound code really shouldn't be digging around inside actors.

Comment (reaperrr) [inline OpenRA.Game/Sound/Sound.cs]: Done.

**Comment by reaperrr** [comment]
Updated.

File: OpenRA.Game/WorldUtils.cs [hunk 7]
```diff
@@ -44,8 +44,8 @@ public static IEnumerable<Actor> FindActorsInCircle(this World world, WPos origi
 
 		public static bool HasVoices(this Actor a)
 		{
-			var selectable = a.Info.Traits.GetOrDefault<SelectableInfo>();
-			return selectable != null && selectable.Voice != null;
+			var voice = a.TraitOrDefault<IVoiced>();
```
Comment (penev92) [inline OpenRA.Game/WorldUtils.cs]: Let's not assume the actor has only one such trait and keep to
`var voice = a.TraitsImplementing<IVoiced>().FirstOrDefault();`

File: OpenRA.Game/WorldUtils.cs [hunk 8]
```diff
@@ -56,9 +56,11 @@ public static bool HasVoice(this Actor a, string voice)
 
 		public static SoundInfo GetVoices(this Actor a)
 		{
-			var selectable = a.Info.Traits.GetOrDefault<SelectableInfo>();
-			if (selectable == null) return null;
-			var v = selectable.Voice;
+			var voiced = a.TraitOrDefault<IVoiced>();
```
Comment (penev92) [inline OpenRA.Game/WorldUtils.cs]: Same here.

File: OpenRA.Game/WorldUtils.cs [hunk 9]
```diff
@@ -44,8 +44,8 @@ public static IEnumerable<Actor> FindActorsInCircle(this World world, WPos origi
 
 		public static bool HasVoices(this Actor a)
```
Comment (pchote) [inline OpenRA.Game/WorldUtils.cs]: It should be relatively simple to remove these three now, but this might be scope creeping to much.

Comment (penev92) [inline OpenRA.Game/WorldUtils.cs]: I was going to object, but I see these three methods have a total of 2 usages, so I'm, inclined to let it go.

Comment (Phrohdoh) [inline OpenRA.Game/WorldUtils.cs]: My opinion is that this is creep and should be done another time.

Comment (reaperrr) [inline OpenRA.Game/WorldUtils.cs]: > It should be relatively simple to remove these three now, but this might be scope creeping to much.

Done

**Comment by reaperrr** [comment]
Updated.

File: OpenRA.Game/Selection.cs [hunk 10]
```diff
@@ -57,9 +57,10 @@ public void Combine(World world, IEnumerable<Actor> newSelection, bool isCombine
 				}
 			}
 
-			var voicedUnit = actors.FirstOrDefault(a => a.Owner == world.LocalPlayer && a.IsInWorld && a.HasVoices());
-			if (voicedUnit != null)
-				Sound.PlayVoice("Select", voicedUnit, voicedUnit.Owner.Country.Race);
+			var voicedActor = actors.FirstOrDefault(a => a.Owner == world.LocalPlayer && a.IsInWorld && (a.TraitsImplementing<IVoiced>() != null));
```
Comment (penev92) [inline OpenRA.Game/Selection.cs]: `.TraitsImplementing()` should never be able to return `null`. The proper check here is probably `.TraitsImplementing().Any()`.
But I suggest using `.HasTrait<IVoiced>()`, which is much simpler ;)

File: OpenRA.Game/WorldUtils.cs [hunk 11]
```diff
@@ -73,10 +53,15 @@ public static void PlayVoiceForOrders(this World w, Order[] orders)
 				if (o.Subject.Destroyed)
 					continue;
 
+				var hasVoice = o.Subject.TraitsImplementing<IVoiced>();
+				if (hasVoice == null)
```
Comment (penev92) [inline OpenRA.Game/WorldUtils.cs]: Suggestion: rename `hasVoice` to `voices` here and use it instead of doing another `o.Subject.TraitsImplementing<IVoiced>()` in the `foreach` 3 lines below.

File: OpenRA.Mods.Common/Traits/Cargo.cs [hunk 12]
```diff
@@ -207,7 +207,7 @@ public string VoicePhraseForOrder(Actor self, Order order)
 			if (order.OrderString != "Unload" || IsEmpty(self))
 				return null;
 
-			return self.HasVoice("Unload") ? "Unload" : "Move";
+			return self.TraitsImplementing<IVoiced>().FirstOrDefault().HasVoice(self, "Unload") ? "Unload" : "Move";
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Cargo.cs]: Please change this to
```
return self.TraitsImplementing<IVoiced>().Any(x => x.HasVoice(self, "Unload")) ? "Unload" : "Move";
```
`.FirstOrDefault().HasVoice()` is a potential NullReferenceException.

File: OpenRA.Mods.Common/Traits/Voiced.cs [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+			var mi = voicedActor.TraitOrDefault<IVoiced>();
+			if (mi == null || mi.VoiceSet == null)
+				return false;
+
+			var type = mi.VoiceSet.ToLowerInvariant();
+			return Sound.PlayPredefined(voicedActor.World.Map.Rules, null, voicedActor, type, phrase, variant, false, pos, volume, true);
+		}
+
+		public bool HasVoices(Actor actor)
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Voiced.cs]: You can delete this method. Nothing is using it any more and there isn't a point to checking if the actor has trait X inside trait X.

File: OpenRA.Mods.Common/Traits/Voiced.cs [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+			return voice != null && voice.VoiceSet != null;
+		}
+
+		public bool HasVoice(Actor actor, string voice)
+		{
+			var v = GetVoices(actor);
+			return v != null && v.Voices.ContainsKey(voice);
+		}
+
+		public SoundInfo GetVoices(Actor actor)
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Voiced.cs]: This method and the above can be simplified to a single
```
		public bool HasVoice(Actor self, string voice)
		{
			if (string.IsNullOrEmpty(Info.VoiceSet))
				return false;

			var voices = self.World.Map.Rules.Voices[Info.VoiceSet.ToLowerInvariant()];
			return voices != null && voices.Voices.ContainsKey(voice);
		}
```

File: OpenRA.Mods.Common/Traits/Voiced.cs [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+		public object Create(ActorInitializer init) { return new Voiced(init.Self, this); }
+	}
+
+	public class Voiced : IVoiced
+	{
+		public readonly VoicedInfo Info;
+
+		public Voiced(Actor self, VoicedInfo info)
+		{
+			Info = info;
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Voiced.cs]: You can get rid of the actors passed as arguments to all the methods if you cache `self` here and use that.

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Voiced.cs]: Convention is to have the first argument to public methods be `Actor self`.

File: OpenRA.Game/Traits/TraitsInterfaces.cs [hunk 16]
```diff
@@ -109,6 +109,15 @@ public interface IDisableMove { bool MoveDisabled(Actor self); }
 
 	public interface ISeedableResource { void Seed(Actor self); }
 
+	public interface IVoiced
+	{
+		bool PlayVoice(string phrase, Actor voicedActor, string variant);
+		bool PlayVoiceLocal(string phrase, Actor voicedActor, string variant, WPos pos, float volume);
+		bool HasVoices(Actor actor);
+		bool HasVoice(Actor actor, string voice);
```
Comment (penev92) [inline OpenRA.Game/Traits/TraitsInterfaces.cs]: Once you address my comments on `Voiced` this whould only have `bool HasVoice(string voice);`

File: OpenRA.Game/Traits/TraitsInterfaces.cs [hunk 17]
```diff
@@ -109,6 +109,15 @@ public interface IDisableMove { bool MoveDisabled(Actor self); }
...
+		string VoiceSet { get; }
```
Comment (penev92) [inline OpenRA.Game/Traits/TraitsInterfaces.cs]: Please move this property before the method declarations (conventions state properties go before methods).

File: OpenRA.Mods.Common/Traits/Voiced.cs [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+
+			var mi = voicedActor.TraitOrDefault<IVoiced>();
+			if (mi == null || mi.VoiceSet == null)
+				return false;
+
+			var type = mi.VoiceSet.ToLowerInvariant();
+			return Sound.PlayPredefined(voicedActor.World.Map.Rules, null, voicedActor, type, phrase, variant, true, WPos.Zero, 1f, true);
+		}
+
+		public bool PlayVoiceLocal(string phrase, Actor voicedActor, string variant, WPos pos, float volume)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Voiced.cs]: * Is there any reason why `pos` wouldn't be the actor position?
* Please rename `voicedActor` to `self` and move to the first arg.
* `self` will not be null, and `mi == this`, so you can make some cleanups in the method body.

File: OpenRA.Mods.Common/Traits/Voiced.cs [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+		{
+			Info = info;
+		}
+
+		public bool PlayVoice(string phrase, Actor voicedActor, string variant)
+		{
+			if (voicedActor == null || phrase == null)
+				return false;
+
+			var mi = voicedActor.TraitOrDefault<IVoiced>();
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Voiced.cs]: Likewise here, `mi == this` and `voicedActor` &rarr; `self`.

File: OpenRA.Mods.Common/Traits/Voiced.cs [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+		public bool PlayVoiceLocal(string phrase, Actor voicedActor, string variant, WPos pos, float volume)
+		{
+			if (voicedActor == null || phrase == null)
+				return false;
+
+			var mi = voicedActor.TraitOrDefault<IVoiced>();
+			if (mi == null || mi.VoiceSet == null)
+				return false;
+
+			var type = mi.VoiceSet.ToLowerInvariant();
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Voiced.cs]: You can replace the last 4 lines with
```
if (string.IsNullOrEmpty(Info.VoiceSet))
    return false;
var type = Info.VoiceSet.ToLowerInvariant();
```

File: OpenRA.Game/Traits/TraitsInterfaces.cs (hunk 16 above)
Comment (Phrohdoh) [inline OpenRA.Game/Traits/TraitsInterfaces.cs]: Per @pchote's comment; keep this as-is.

Comment (penev92) [inline OpenRA.Game/Traits/TraitsInterfaces.cs]: Actually, still remove `HasVoices()` and chenge this to `Actor self`.

File: OpenRA.Mods.Common/Traits/Voiced.cs (hunk 18 above)
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Voiced.cs]: > Is there any reason why pos wouldn't be the actor position?

Not really, that was just copy-pasta.

**Comment by reaperrr** [comment]
Updated.
Adressed everything that was pointed out, I think.

File: OpenRA.Game/WorldUtils.cs [hunk 21]
```diff
@@ -70,13 +50,19 @@ public static void PlayVoiceForOrders(this World w, Order[] orders)
 				if (o == null)
 					continue;
 
-				if (o.Subject.Destroyed)
+				var orderSubject = o.Subject;
+				if (orderSubject.Destroyed)
+					continue;
+
+				var voices = orderSubject.TraitsImplementing<IVoiced>();
+				if (voices == null)
```
Comment (penev92) [inline OpenRA.Game/WorldUtils.cs]: `if (!voices.Any())`

File: OpenRA.Mods.Common/Traits/Sound/DeathSounds.cs [hunk 22]
```diff
@@ -40,11 +40,11 @@ public void Killed(Actor self, AttackInfo e)
 			if (e.Warhead == null)
 				return;
 
-			var cp = self.CenterPosition;
-
-			if (info.DeathTypes.Contains(e.Warhead.DeathType) ||
-				(!info.DeathTypes.Any() && !self.Info.Traits.WithInterface<DeathSoundsInfo>().Any(dsi => dsi.DeathTypes.Contains(e.Warhead.DeathType))))
-				Sound.PlayVoiceLocal(info.DeathSound, self, self.Owner.Country.Race, cp, info.VolumeMultiplier);
+			foreach (var voiced in self.TraitsImplementing<IVoiced>())
```
Comment (penev92) [inline OpenRA.Mods.Common/Traits/Sound/DeathSounds.cs]: Please move the `foreach` inside the `if`.

**Comment by penev92** [comment]
Code looks good and everything sounds good in all 4 mods.
:+1: 
Also a very nice cleanup.

**Comment by reaperrr** [comment]
Should I squash some (or all) of the commits before this is merged?

**Comment by abcdefg30** [comment]
Yeah, I think squashing some of the commits together is a good idea, but you don't need to.

**Comment by reaperrr** [comment]
Squashed some of the commits and (finally) added a fix that closes #8121.

File: OpenRA.Mods.Common/Traits/Cargo.cs [hunk 23]
```diff
@@ -207,7 +207,7 @@ public string VoicePhraseForOrder(Actor self, Order order)
 			if (order.OrderString != "Unload" || IsEmpty(self))
 				return null;
 
-			return self.HasVoice("Unload") ? "Unload" : "Move";
+			return self.TraitsImplementing<IVoiced>().Any(x => x.HasVoice(self, "Unload")) ? "Unload" : "Move";
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Cargo.cs]: This is the only trait that hardcodes a fallback to "Move".  How about we fix the voice sets and remove this?

**Comment by pchote** [comment]
Having the `foreach (var voiced in self.TraitsImplementing<IVoiced>())` everywhere feels a bit ugly.  Could you add a `PlayVoice` (and maybe also `HasVoice`?) extension method to ActorExts, and then use `actor.PlayVoice` instead?

**Comment by reaperrr** [comment]
Updated.

**Comment by reaperrr** [comment]
Updated.

**Comment by abcdefg30** [comment]
Sorry, but this needs a rebase again.

File: OpenRA.Mods.Common/ActorExts.cs [hunk 24]
```diff
@@ -80,6 +80,39 @@ public static Target ResolveFrozenActorOrder(this Actor self, Order order, Color
 			return Target.Invalid;
 		}
 
+		public static void PlayVoice(this Actor self, Actor actor, string phrase, string variant)
```
Comment (pchote) [inline OpenRA.Mods.Common/ActorExts.cs]: The `actor` parameters here and below aren't used (and don't make a lot of sense?).  Could you please remove them?

File: OpenRA.Game/WorldUtils.cs [hunk 25]
```diff
@@ -70,13 +50,15 @@ public static void PlayVoiceForOrders(this World w, Order[] orders)
...
-				if (o.Subject.Destroyed)
+				var orderSubject = o.Subject;
+				if (orderSubject.Destroyed)
 					continue;
 
-				foreach (var v in o.Subject.TraitsImplementing<IOrderVoice>())
-					if (Sound.PlayVoice(v.VoicePhraseForOrder(o.Subject, o),
-						o.Subject, o.Subject.Owner.Country.Race))
-						return;
+				foreach (var voice in orderSubject.TraitsImplementing<IVoiced>())
```
Comment (pchote) [inline OpenRA.Game/WorldUtils.cs]: This should be able to use the new extension methods.

File: OpenRA.Game/Selection.cs [hunk 26]
```diff
@@ -57,9 +57,10 @@ public void Combine(World world, IEnumerable<Actor> newSelection, bool isCombine
 				}
 			}
 
-			var voicedUnit = actors.FirstOrDefault(a => a.Owner == world.LocalPlayer && a.IsInWorld && a.HasVoices());
-			if (voicedUnit != null)
-				Sound.PlayVoice("Select", voicedUnit, voicedUnit.Owner.Country.Race);
+			var voicedActor = actors.FirstOrDefault(a => a.Owner == world.LocalPlayer && a.IsInWorld && a.HasTrait<IVoiced>());
+			if (voicedActor != null)
+				foreach (var voice in voicedActor.TraitsImplementing<IVoiced>())
```
Comment (pchote) [inline OpenRA.Game/Selection.cs]: This should be able to use the new extension methods.

**Comment by pchote** [comment]
Looks like we are converging to a nice clean up here :)  I've pointed out a couple of places you have missed, which should be easy fixups.

File: OpenRA.Mods.Common/ActorExts.cs (hunk 24 above)
Comment (reaperrr) [inline OpenRA.Mods.Common/ActorExts.cs]: I misunderstood something, will fix this.

File: OpenRA.Game/WorldUtils.cs (hunk 25 above)
Comment (reaperrr) [inline OpenRA.Game/WorldUtils.cs]: This and Selection are still in Game, while ActorExts are in Common. I could move the PlayVoice/HasVoice to Game of course (though I don't know into which file), but I believe the long-term goal was rather to move more stuff from the Game engine to Common, right?

**Comment by reaperrr** [comment]
Updated and rebased.

Someone notify me when this is otherwise ready to merge, then I'll squash some commits (unless you think it's fine the way it is).

File: OpenRA.Game/VoiceExts.cs [hunk 27]
```diff
@@ -0,0 +1,74 @@
...
+using System.Collections.Generic;
+using System.Drawing;
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA
+{
+	public static class VoiceExts
+	{
+		public static void PlayVoice(this Actor self, string phrase, string variant)
```
Comment (pchote) [inline OpenRA.Game/VoiceExts.cs]: AFAIK the variant here and below are always just the owner's race.  We can probably drop that argument and simplify this a step further.

Comment (pchote) [inline OpenRA.Game/VoiceExts.cs]: We probably want to keep the variant string on the lower level code for now though, as they are used for different things by e.g. notifications.

File: OpenRA.Mods.Common/Traits/Sound/AnnounceOnBuild.cs [hunk 28]
```diff
@@ -19,7 +19,7 @@ public class AnnounceOnBuild : INotifyBuildComplete
 	{
 		public void BuildingComplete(Actor self)
 		{
-			Sound.PlayVoice("Build", self, self.Owner.Country.Race);
+			self.PlayVoice("Build", self.Owner.Country.Race);
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Traits/Sound/AnnounceOnBuild.cs]: This and the "Kill" below should probably also be moved into the Info class, like the "Unload" above. Doesn't have to be in this PR, though, just sayin'.

**Comment by obrakmann** [comment]
All orders except normal move orders don't generate a response from the receiver of the order. AttackMoves, capturing, entering transports are all silent, both in td and ra, so I assume everywhere.

**Comment by reaperrr** [comment]
Adressed everything that was mentioned.
I had to revert the changes to VoicePhraseForOrder though, because they were responsible for the issue @obrakmann pointed out

**Comment by penev92** [comment]
I've been keeping track of the new changes (thanks for making them in new commits), so my +1 still stands.

File: OpenRA.Game/Traits/Selectable.cs [hunk 29]
```diff
@@ -20,7 +20,6 @@ public class SelectableInfo : ITraitInfo
 		public readonly bool Selectable = true;
```
Comment (Mailaender) [inline OpenRA.Game/Traits/Selectable.cs]: You can remove this now.

Comment (Mailaender) [inline OpenRA.Game/Traits/Selectable.cs]: However this pull request is already large enough.

**Comment by Mailaender** [comment]
:+1: Thanks. This is a valueable cleanup. Works as promised. :white_check_mark:

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/d8cbdb9243e1b61414db4344106be04f8e188a05...d49882200e5762d26480fc40cd7bcae72d165f48)
